### ISSUE TITLE
Fetch:template throws exception when trying to copy a symlink

### ISSUE THREAD

**Issue by pvandervelde** [backstage/backstage#12377]
## Expected Behavior

When a template contains symlinks I expect it to copy the symlink

## Actual Behavior

When the fetch:template action tries to copy a symlink an exception is thrown and Backstage stops processing the template.

## Steps to Reproduce

1. In backstage load a new template using this template: https://github.com/pvandervelde/backstage-fetch-template-symlink-issue/blob/master/symlink-issue/template.yaml
2. Try to use the template, fill out any name
3. Notice that when running the template an error happens when backstage tries to copy the symlink file. 

There are 3 symlink files in the repo. One points to the readme of the repository. One points to a file called `hello.txt` one level up from the repository, and one points to a non-existing file. None of the symlinks can be copied by the templater. It looks like this is an issue caused by `isBinaryFile` here: https://github.com/backstage/backstage/blob/89396a2032caa1ff3b8a7c8b6e3ae416d4c056c7/plugins/scaffolder-backend/src/scaffolder/actions/builtin/fetch/template.ts#L223

The documentation for `isBinaryFile` indicates that the function will throw an exception if it cannot open the file to find out if the file is binary or not.

## Context

We are using Backstage templates to create new service repositories. As part of this process we also generate PRs for an IaC repository. This IaC repository is a monorepo that uses symlinks to link to common build files. In order for us to create these files
we either need to create them when the PR is created (which isn't possible as far as I know) or we pre-create them and then just copy them (which is what we're trying). The layout of the IaC repository is something like this

```
iac_repo
    [D] env
        [D] prod
            [D] services
                [D] service_1
                    [symlink] Makefile
                    [symlink][D] common
                    main.tf
                    ... other terraform files specific to the service ...
                [D] service_2
                    ...
                [D] service_3
                    ...
                ...
    [D] common
        .. shared terraform files ...
    Makefile
```

As part of the template we want to be able to add the `service_N` directories using a PR to the repository. For that we'll need to add the symlinks.

Steps we take in our template:

1. Fetch the contents of the repository and put it through the templater
2. Create a new repository for the service and fill it with the content that was just templated
3. Fetch the scripts and configs for our the IaC repo. One of the files is a symlink that will point to the Make file in the IaC repo
4. Create a PR with the scripts and configs we just fetched

<img width="1539" alt="image" src="https://user-images.githubusercontent.com/96201/176598102-8976156c-8a8c-4828-8756-f88bec0eae23.png">

## Your Environment

- Browser Information: Google Chrome 103.0.5060.53 (Official Build) (x86_64)

- Output of `yarn backstage-cli info`: 

```text
OS:   Darwin 21.5.0 - darwin/x64
node: v16.15.1
yarn: 1.22.19
cli:  0.17.2 (installed)

Dependencies:
  @backstage/app-defaults                          1.0.3
  @backstage/backend-common                        0.14.0
  @backstage/backend-tasks                         0.3.2
  @backstage/backend-test-utils                    0.1.25
  @backstage/catalog-client                        1.0.3
  @backstage/catalog-model                         1.0.3
  @backstage/cli-common                            0.1.9
  @backstage/cli                                   0.17.2
  @backstage/config-loader                         1.1.2
  @backstage/config                                1.0.1
  @backstage/core-app-api                          1.0.3
  @backstage/core-components                       0.9.5
  @backstage/core-plugin-api                       1.0.3
  @backstage/errors                                1.0.0
  @backstage/integration-react                     1.1.1
  @backstage/integration                           1.2.1
  @backstage/plugin-api-docs                       0.8.6
  @backstage/plugin-app-backend                    0.3.33
  @backstage/plugin-auth-backend                   0.14.1
  @backstage/plugin-auth-node                      0.2.2
  @backstage/plugin-bazaar-backend                 0.1.17
  @backstage/plugin-bazaar                         0.1.21
  @backstage/plugin-catalog-backend-module-aws     0.1.6
  @backstage/plugin-catalog-backend-module-github  0.1.4
  @backstage/plugin-catalog-backend                1.2.0
  @backstage/plugin-catalog-common                 1.0.3
  @backstage/plugin-catalog-graph                  0.2.18
  @backstage/plugin-catalog-import                 0.8.9
  @backstage/plugin-catalog-react                  1.1.1
  @backstage/plugin-catalog                        1.3.0
  @backstage/plugin-cicd-statistics                0.1.8
  @backstage/plugin-circleci                       0.3.6
  @backstage/plugin-cost-insights                  0.11.28
  @backstage/plugin-explore-react                  0.0.18
  @backstage/plugin-explore                        0.3.37
  @backstage/plugin-github-actions                 0.5.6
  @backstage/plugin-github-pull-requests-board     0.1.0
  @backstage/plugin-home                           0.4.22
  @backstage/plugin-kafka-backend                  0.2.26
  @backstage/plugin-kafka                          0.3.6
  @backstage/plugin-org                            0.5.6
  @backstage/plugin-pagerduty                      0.4.0
  @backstage/plugin-permission-common              0.6.2
  @backstage/plugin-permission-node                0.6.2
  @backstage/plugin-permission-react               0.4.2
  @backstage/plugin-proxy-backend                  0.2.27
  @backstage/plugin-scaffolder-backend             1.3.0
  @backstage/plugin-scaffolder-common              1.1.1
  @backstage/plugin-scaffolder                     1.3.0
  @backstage/plugin-search-backend-module-pg       0.3.4
  @backstage/plugin-search-backend-node            0.6.2
  @backstage/plugin-search-backend                 0.5.3
  @backstage/plugin-search-common                  0.3.4, 0.3.5
  @backstage/plugin-search-react                   0.2.1
  @backstage/plugin-search                         0.9.0
  @backstage/plugin-stack-overflow                 0.1.2
  @backstage/plugin-tech-radar                     0.5.13
  @backstage/plugin-techdocs-backend               1.1.2
  @backstage/plugin-techdocs-module-addons-contrib 1.0.1
  @backstage/plugin-techdocs-node                  1.1.2
  @backstage/plugin-techdocs-react                 1.0.1
  @backstage/plugin-techdocs                       1.2.0
  @backstage/plugin-user-settings                  0.4.5
  @backstage/release-manifests                     0.0.4
  @backstage/search-common                         0.3.4
  @backstage/test-utils                            1.1.1
  @backstage/theme                                 0.2.15
  @backstage/types                                 1.0.0
  @backstage/version-bridge                        1.0.1
```

**Comment by benjdlambert** [comment]
@pvandervelde, so I did some digging I think it should work for files that do exist in the template. But are you saying that it's not for you right now? The `isBinaryFile` will try to resolve the `realpath` so I guess is what happening is that if the file that the symlink is referring to doesn't exist then it's gonna throw out.

https://github.com/benjdlambert/backstage-fetch-template-symlink-issue I made it work for the link thats in the repo, but not that's not in the repo.

I wonder if we can treat symlinks differently, and try to resolve them to see if they point to something that should be in the repo and then just copy them anyway providing they reference something inside the workspace and skip the `isBinaryFile` check.

We want to resolve the `realpath` of the symlink to make sure that h4xx0rs can't reference things like `/etc/passwd` or whatever. Do you think this would solve the issue for you?

**Comment by marcus-crane** [comment]
@benjdlambert Thanks for having a look 🙂 

I think the issue we were facing is that `isBinaryFile` would fail (which is fine) but then the [fs.stat check](https://github.com/backstage/backstage/blob/89396a2032caa1ff3b8a7c8b6e3ae416d4c056c7/plugins/scaffolder-backend/src/scaffolder/actions/builtin/fetch/template.ts#L229) would try to resolve the link, which points at something that doesn't exist (yet) and fail

Treating the symlink as a binary should indeed fix this I tihnk

**Comment by benjdlambert** [comment]
This actually got released today in v1.3.1 so you should be able to bump the `scaffolder-backend` dependency in your backend to `1.3.1` and it should just work ™️ for you.
